### ISSUE TITLE
Game freezes when changing animations on an animator with a large number of drives

### ISSUE THREAD

**Issue by Foxxboxx** [Yellow-Dog-Man/Resonite-Issues#2699]
### Describe the bug?

When an animation on an animator is changed that contains a large number of drives the game will freeze for up to 3 seconds. @bontebok and @stiefeljackal conducted a regression test and confirmed that the issue was first observed in version 2024.7.11.1293. This may be related to the new Type Management System that was introduced in the previous version. 

@bontebok will upload video and logs for the steps taken to reproduce the issue for versions 2024.7.4.1347 and 2024.7.11.1293
@stiefeljackal will upload video and logs for the steps taken to reproduce the issue for version 2024.7.10.1399

### To Reproduce

1. Launch Resonite as normal.
2. Go to the CJ: New Years Eve 2024 world. resrec:///G-Creator-Jam/R-ef0b9779-44c5-4da1-bb8f-9facf83138fe
3. Go to the Drone Show control panel in the world and start the show.
4. The first freeze will occur when the show transitions from the count down animation to the main show animation. The game will freeze for all connected clients for up to 3 seconds. 
5. Fast forward or watch the main show until the end. When the show transitions back to the start of the count down to show the drones in the air, the game will freeze for up to 3 seconds again. 

Repeating steps 4 and 5 continues to reproduce the freeze. 

### Expected behavior

Changing the animation on an animator should not cause the game to freeze. This did not occur in this world during the NYE Party for 2024. 

### Resonite Version Number

2024.7.25.1284

### What Platforms does this occur on?

Windows

### What headset if any do you use?

Oculus Quest 2

### Log Files

[DESKTOP-UQPLP8K - 2024.7.25.1284 - 2024-07-31 19_49_13.log](https://github.com/user-attachments/files/16449398/DESKTOP-UQPLP8K.-.2024.7.25.1284.-.2024-07-31.19_49_13.log)

### Reporters

foxbox. rucio stiefeljackal

**Comment by bontebok** [comment]
As promised, here are video clips and logs demonstrating performing the steps to reproduce the issue in 2024.7.4.1347 and 2024.7.11.1293. As you can see, the freezing behavior is not exhibited in 2024.7.4.1347 but does appear in 2024.7.11.1293.


https://github.com/user-attachments/assets/094857a1-19a8-4622-8756-08a92be04219

[DESKTOP-M4CKTHS - 2024.7.4.1347 - 2024-07-31 21_49_58.log](https://github.com/user-attachments/files/16449418/DESKTOP-M4CKTHS.-.2024.7.4.1347.-.2024-07-31.21_49_58.log)

https://github.com/user-attachments/assets/4ddfbbfb-e3b8-4c69-9d22-c54cf400d8cb

[DESKTOP-M4CKTHS - 2024.7.11.1293 - 2024-07-31 21_43_08.log](https://github.com/user-attachments/files/16449419/DESKTOP-M4CKTHS.-.2024.7.11.1293.-.2024-07-31.21_43_08.log)

**Comment by stiefeljackal** [comment]
Also as promised, here is the video and log that performed the same steps to reproduce the issue in 2024.7.10.1399. In that version, the freezing behavior is not exhibited at all.


https://github.com/user-attachments/assets/a5565e50-14ce-4520-9de1-0d7cac7460ed


https://github.com/user-attachments/assets/9a48e2f5-1a93-46c4-92d5-a0803e7c96ac
[JACKAL_MK2 - 2024.7.10.1399 - 2024-07-31 22_49_55.log](https://github.com/user-attachments/files/16449622/JACKAL_MK2.-.2024.7.10.1399.-.2024-07-31.22_49_55.log)

**Comment by shiftyscales** [comment]
Seeking input from @Frooxius.

**Comment by stiefeljackal** [comment]
> How many fields are being driven by this?

There are 2000 fields being driven. Breaking it down, 1000 are on a `float3` field (Position) and 1000 are on a `colorX` field.

**Comment by stiefeljackal** [comment]
> The new type system might've introduced more overhead for certain operations with the data model, that don't play with how the Animator currently works.

@Frooxius That is what we suspected as well as pinpointed that as the cause.

My video demonstrates the animation without any freezing on a client from the `2024.7.10.1399` build, which is the last version before the new type management system.

@bontebok's second video demonstrates the freezing that occurs with the animation on a client from the `2024.7.11.1293` build, which is the first version the type management system was introduced.

This freezing issue still occurs on the latest version of Resonite (`2024.7.25.1284` at the time of writing).

**Comment by knackrack615** [comment]
Can confirm this only started happening after the new type system was introduced, as a temporary workaround [this mod](https://github.com/BlueCyro/EffortlessEfficiency) by Cyro fixes it.

I am driving a single Animator between 3 different animations (the funni cat next to my avi, changes animation based on my movement)

With EffortlessEfficiency mod

https://github.com/user-attachments/assets/c51d0734-38ed-479f-b876-de3f55b8b964

Without mods

https://github.com/user-attachments/assets/aced1633-9543-4fd1-8bb8-b52e22992d7f

**Comment by Frooxius** [comment]
@stiefeljackal Thanks for the info!

Yes, I was actually reacting to the statements that the type system might've introduced this problem and adding on why.

We haven't taken any steps to fix this yet, so we wouldn't expect this problem to disappear until then.

**Comment by stiefeljackal** [comment]
> Can confirm this only started happening after the new type system was introduced, as a temporary workaround [this mod](https://github.com/BlueCyro/EffortlessEfficiency) by Cyro fixes it.

@knackrack615 I'm curious. Can you try this with the CJ: New Years Eve 2024 world? The `resrec` for it is `resrec:///G-Creator-Jam/R-ef0b9779-44c5-4da1-bb8f-9facf83138fe`. The player for the show is located on the middle leaf of the structure.

**Comment by knackrack615** [comment]
> > Can confirm this only started happening after the new type system was introduced, as a temporary workaround [this mod](https://github.com/BlueCyro/EffortlessEfficiency) by Cyro fixes it.
> 
> @knackrack615 I'm curious. Can you try this with the CJ: New Years Eve 2024 world? The `resrec` for it is `resrec:///G-Creator-Jam/R-ef0b9779-44c5-4da1-bb8f-9facf83138fe`. The player for the show is located on the middle leaf of the structure.

Can confirm it does indeed work fine with EffortlessEfficiency loaded

https://github.com/user-attachments/assets/6f37c833-f557-4c39-a1e2-4ca9e3123883

**Comment by Frooxius** [comment]
That sounds like a good lead with the reflection possibly being the caused.

One thing that I noticed (but haven't confirmed), there seems to be an issue with reflection in Mono, where heavy multi-threaded reflection use during startup seems to make other things slower. When I run Resonite for the first time with new build, when the startup generates compatibility hashes, things seem to lag more for that run (e.g. opening new UI's). On subsequent runs it's better.

Would you be willing to do a test with this? It's not super necessary, but would give us a bit more data.

Essentially run Resonite with the assembly metadata cache cleared, do the test, see if it lags longer than on the subsequent startup.

**Comment by ohzee00** [comment]
I noticed something similar when testing this issue:
https://github.com/Yellow-Dog-Man/Resonite-Issues/issues/2671

In the Additional Context section I noted that it feels (as in, not exact numbers) that on first startup without assembly metadata it feels like large UI pieces like Worlds or mostly Contacts stutter a lot more.

**Comment by knackrack615** [comment]
> That sounds like a good lead with the reflection possibly being the caused.
> 
> One thing that I noticed (but haven't confirmed), there seems to be an issue with reflection in Mono, where heavy multi-threaded reflection use during startup seems to make other things slower. When I run Resonite for the first time with new build, when the startup generates compatibility hashes, things seem to lag more for that run (e.g. opening new UI's). On subsequent runs it's better.
> 
> Would you be willing to do a test with this? It's not super necessary, but would give us a bit more data.
> 
> Essentially run Resonite with the assembly metadata cache cleared, do the test, see if it lags longer than on the subsequent startup.

After some testing I can confirm that deleting assembly metadata cache makes the game take significantly longer to start (In my case with my mods we're talking about going from around a minute to 4+)

On unmodded, the game takes about 1m50s, with the cache built it takes around 45s

**Comment by ohzee00** [comment]
> After some testing I can confirm that deleting assembly metadata cache makes the game take significantly longer to start (In my case with my mods we're talking about going from around a minute to 4+)
> 
> On unmodded, the game takes about 1m50s, with the cache built it takes around 45s

I believe the request was more so the lag _after_ the metadata cache is generated for the first time, not start up time. It is already known that startup time can drastically change based on cache or lack thereof.

**Comment by knackrack615** [comment]
> > After some testing I can confirm that deleting assembly metadata cache makes the game take significantly longer to start (In my case with my mods we're talking about going from around a minute to 4+)
> > On unmodded, the game takes about 1m50s, with the cache built it takes around 45s
> 
> I believe the request was more so the lag _after_ the metadata cache is generated for the first time, not start up time. It is already known that startup time can drastically change based on cache or lack thereof.

You are absolutely right, I completely misread that, yeah things seem to be _a bit_ worse on first start, but not by that much on my testing (Talking about ~1-2s worth of extra stuttering in my case)

**Comment by Foxxboxx** [comment]
I did some more testing with the animations in this world. I ran a test on the current mono runtime build and the new .NET 8.0 build for the headless and monitored the headless performance. The issue is exhibited in the mono runtime build but is not exhibited in the .NET 8.0 build. I tested by spamming "users" into the console on both builds to see if I could see any FPS drop and monitored a panel in world that shows the FPS of the headless. Attached are videos of the in world experiment. 

Mono Build that shows FPS drop for the headless when animations are switched.
https://github.com/user-attachments/assets/67e0df1d-e56c-49d9-995a-fc568fbffe24

.NET 8.0 Build, no FPS drop detectable when switching animations. 
https://github.com/user-attachments/assets/136ae1f9-fc4d-45cc-8b09-72dac3af0579

**Comment by Frooxius** [comment]
Thank you for the info! The .NET 8.0 test helps there too.

I've noticed that Reflection under Mono is significantly slower than .NET 8.0.

Question there is, do we spend time on this issue specifically right now, or do we just head for moving the graphical client to .NET 8.0? That would likely resolve the issue implicitly, but it depends on how big of a problem is it right now.

**Comment by Foxxboxx** [comment]
I'll be honest. I'm bias toward fixing it now rather than later. Partly because myself and the rest of the Slerps worked hard on this world to make it a beautiful and polished example of not only what we can do, but what the platform can do. The second is, as a mentor, I often show off this world to new users so they can see what the platform is capable of. And having it freeze during the transitions detracks from that. 

But its up to you all, those are just my thoughts and opinions.

**Comment by Frooxius** [comment]
We could fix it now. Especially if this comes up often.

The only flip side is that it will delay other work - particularly the switch to .NET 8.0 for graphical client, which could fix in a few months time anyway, "wasting" the time we spend on it now, so that's the calculus I'm essentially making: Is this a big enough issue to fix this specifically, to delay other performance improvements that affect everything a bit?

**Comment by bontebok** [comment]
> We could fix it now. Especially if this comes up often.
> 
> The only flip side is that it will delay other work - particularly the switch to .NET 8.0 for graphical client, which could fix in a few months time anyway, "wasting" the time we spend on it now, so that's the calculus I'm essentially making: Is this a big enough issue to fix this specifically, to delay other performance improvements that affect everything a bit?

Adding my 2 cents here. I agree with FoxBox that it would be nice if this was resolved to show off the NYE world, but as we have already started planning for the NYE show for 2025 and we want to continue with the tradition of creating a drone show, this would have the largest impact to us on January 1st.

I reviewed @BlueCyro's EffortlessEfficiency Mod which implements a very simple but effective solution that, although does not fix the problem, will mask the issue entirely for users causing a significant speed up when switching the animator references.

Would you be willing to review the caching solution from @BlueCyro's EffortlessEfficiency patch and consider implementing it in the Mono client to hold us over from now until New Year's, or the release of the .NET 8 graphical client (whichever comes first)? I think this should be a low-time/low-effort solution and not take away from other important performance and graphical client priorities.

Thanks!

**Comment by Frooxius** [comment]
I don't think this would be an issue for NYE show for 2025 - I think there's a good likelyhood the graphical client will be moved to .NET 8.0 at that point.

I've looked at the patch, but I'll need to spend more time on it to make sure it doesn't introduce additional problems - that's the main thing I'm trying to determine now - should I invest time into this right now? It might end up being quicker to fix, but I'm trying to gauge is how urgent it is right now and how much is affected by it.

**Comment by bontebok** [comment]
It's not urgent for me, but January tends to sneak up fast towards the end of the year. If blockers prevent the move to the graphical client by New Years, it might become urgent. I'm good with putting a hold on this and check back around 12/1 and see how things are looking for New Years. Will that help alleviate the pressure and let the team focus on migrating to .NET 8 and the new graphical client?

**Comment by Frooxius** [comment]
Yeah we could do that. If it looks like the graphical client on .NET 8.0 isn't going to pan out for New Years, we can pause it and prioritize this one at that point.

If it ends up panning out, then it'll just be resolved implicitly.

If there are cases where this is urgent right now, we could still prioritize it right now.

**Comment by Foxxboxx** [comment]
In my opinion I would like the .NET 8 build to have as minimum delays as possible. The performance we've seen on the headless has been insane and I think getting the graphical client there would be an insane improvement for the user experience. If it was out before NYE I can only imagine the potential we could have for a giant party. While last year we had 5 or 6bworlds capped at 20 users. I can see ourselves having 50+ user sessions.  Which is just insane to think of. 

However, the drone show we do every year is one of the big focal points. And I agree with Rucio, that I would want the animator, or any underlying system, fixed prior to NYE 2025. So we can put on an amazing show for everyone. Even if we aren't ready for .NET 8 yet. So having one or the other would be very much appreciated before then. Other than that I can wait.

**Comment by stiefeljackal** [comment]
My initial thought is to wait for the .NET client to be released as that is the most important piece in improving performance, and that is still my thought. There isn't an urgency to have a solution in place right now, and time and energy should be focused on the new .NET client. As @bontebok has mentioned, this issue can be revisited in early December if the newer client is not released by then.

**Comment by BlueCyro** [comment]
> Yeah we could do that. If it looks like the graphical client on .NET 8.0 isn't going to pan out for New Years, we can pause it and prioritize this one at that point.
> 
> If it ends up panning out, then it'll just be resolved implicitly.
> 
> If there are cases where this is urgent right now, we could still prioritize it right now.

If it'd take some pressure off of you, I could slap this in real quick since I've done it before

**Comment by bontebok** [comment]
> Yeah we could do that. If it looks like the graphical client on .NET 8.0 isn't going to pan out for New Years, we can pause it and prioritize this one at that point.
> 
> If it ends up panning out, then it'll just be resolved implicitly.
> 
> If there are cases where this is urgent right now, we could still prioritize it right now.

Hi Froox, sorry for the ping on this - we're in the middle of NYE development and wondering if you're still planning to deploy a fix before New Years Eve or if we should plan around the issue.

Thanks!

**Comment by Stellanora64** [comment]
@BlueCyro do you think you could look into this?

**Comment by Frooxius** [comment]
@bontebok @Stellanora64 I'm sorry for lack of response, I missed your original message on this.

We'll sort it out and push a build with a workaround fix in the next few days so you don't have to worry about it.

**Comment by Frooxius** [comment]
This should be fixed in 2024.12.26.1378 now! Can you please give it a test and see if it's okay? The freeze doesn't happen anymore.

Sorry for the wait, things are a bit busy with holiday stuff.

**Comment by knackrack615** [comment]
All looks good here! Zero freezing/performance drops.

https://github.com/user-attachments/assets/c304401b-76cd-4e1a-9be6-b8f4e8e130ee

Thank you Froox!

**Comment by bontebok** [comment]
I confirmed from my laptop that without Cyro's mod the animation transitions occur with no freeze, thank you! Did you go with a caching approach like @BlueCyro's mod or did you find another solution?

**Comment by Frooxius** [comment]
Glad it helps!

Yeah I used the caching approach. I discussed it with @BlueCyro a bit too in person too (we're at the same house for Holidays). 

There seems to be a weird issue with how Mono implements reflection, where the more it gets used, the slower any subsequent calls to it get. Since the new Type system hits reflection a lot more during startup, it somehow makes reflection heavy stuff a lot worse than it was.
